### PR TITLE
Add postgres profile

### PR DESCRIPTION
**Pull request by JamesChristie**
This is the addition of a profile to utilize the changes being made to this [package](https://github.com/whyrusleeping/sql-datastore/pull/2).

### CODE REVIEW

File: profile.go [hunk 1]
```diff
@@ -126,6 +126,25 @@ Make sure to backup your data frequently.`,
...
+WARNING: postgres datastore is experimental.
+Make sure to backup your data frequently.`,
+		Transform: func(c *Config) error {
+			c.Datastore.Spec = map[string]interface{}{
+				"type":   "measure",
+				"prefix": "postgres.datastore",
+				"child": map[string]interface{}{
+					"type":       "postgres",
+					"path":       "postgres",
+					"syncWrites": true,
```
Comment (kevina) [inline profile.go]: Does the "syncWrites" option mean anything to postgres?

Comment (magik6k) [reply profile.go]: Looks like it's not handled in https://github.com/ipfs/go-ipfs/pull/5403/files either, so this can be just dropped

Comment (i-norden) [reply profile.go]: Dropped the syncWrites an updated go-ipfs-addr and go-libp2p-peer dependencies. Can this be merged?

**Comment by i-norden** [comment]
Sorry for the git mess, tried to clean it up a bit but can't seem to get rid of ab850a5 on this fork. I did not do a gx release this time since I moved away from head state to try and clean up the git history. Is it okay if we do that after we merge? I will then update that gx hash in the go-ipfs PR (and iptb-plugins).

**Comment by i-norden** [comment]
The issue with ab850a5 appears to be a bug in how GitHub shows the order of commits as discussed [here](https://help.github.com/articles/why-are-my-commits-in-the-wrong-order/) [here](https://github.com/isaacs/github/issues/974) and [here](https://github.com/isaacs/github/issues/386). The correct git history is perserved otherwise. I think the only way to fix this on GitHub would be to create a new fork/PR.

**Comment by i-norden** [comment]
Rebasing fixed the git mess, sorry about that!

File: profile.go [hunk 2]
```diff
@@ -131,6 +132,29 @@ Make sure to backup your data frequently.`,
...
+Make sure to backup your data frequently.`,
+		Transform: func(c *Config) error {
+			c.Datastore.Spec = map[string]interface{}{
+				"type":   "measure",
+				"prefix": "postgres.datastore",
+				"child": map[string]interface{}{
+					"type":     "postgres",
+					"host":     os.Getenv("IPFS_PGHOST"),
+					"user":     os.Getenv("IPFS_PGUSER"),
+					"passfile": os.Getenv("IPFS_PGPASSFILE"),
```
Comment (magik6k) [inline profile.go]: `go fmt ./...`

Comment (i-norden) [reply profile.go]: I'm not seeing any changes when I go fmt ./... locally

Comment (magik6k) [reply profile.go]: Hmm.. Probably github just being weird

**Comment by i-norden** [comment]
Lmk if there is anything I can do to help get this merged in, thanks!

**Comment by magik6k** [comment]
@Stebalien ping

**Comment by Stebalien** [comment]
1. Do we want to include postgres support by-default or should we make it a plugin? If we go the plugin route, the plugin could register it's own profile.
2. Using environment variables like this sketches me out a bit. Given all those variables, does this buy anything over manually updating the config? Is this an issue because the `datastore_spec` file makes changing datastores after initialization difficult?

**Comment by magik6k** [comment]
> 1. the plugin could register it's own profile.

:+1:, but I'm not sure how (preloaded) plugins work with init.

> 2. [...] Is this an issue because the datastore_spec file makes changing datastores after initialization difficult?

We write some things to the datastore on init, so we need have a way to access it at init

**Comment by i-norden** [comment]
When this PR was first opened up the Postgres datastore was being built directly into the [fsrepo/datastores.go](https://github.com/ipfs/go-ipfs/blob/master/repo/fsrepo/datastores.go), but then most of the datastores were moved to being plugins and so we followed suite here.  As @magik6k points out, if we don't have the Postgres datastore configured with the necessary variables at `init` we run into an error at that time as it needs to connect and write to the database.

**Comment by Stebalien** [comment]
So, we _can_ do this with preloaded plugins. We load them before calling init so they can add profiles in `Plugin.Init`. Unfortunately, this won't work for non-preloaded plugins. For that, we needhttps://github.com/ipfs/go-ipfs/pull/4244.

**Comment by i-norden** [comment]
Hi all, wanted to check in and see if I could finish up this set of PRs. Not entirely clear to me what else needs to be done. I am personally partial to including the Postgres datastore as one of the preloaded plugins, this makes it straight forward to use Postgres when working with an ipfs node as part of another codebase. 

E.g. just need to do:
```go
l, err := loader.NewPluginLoader("")
	if err != nil {
		return nil, err
	}
	err = l.Initialize()
	if err != nil {
		return nil, err
	}
	err = l.Inject()
	if err != nil {
		return nil, err
	}
	node, err := ipfs.InitIPFSNode(ipfsPath)
	if err != nil {
		return nil, err
	}
```
And it is able to work with the Postgres backed datastore (if that is what was initialized with `ipfs init`)

**Comment by Stebalien** [comment]
Postgres is a large dependency to add by default.

> Its not clear to me exactly how we would use it this way otherwise, I assume we would need to build the plugin "manually" and then link to its path in the loader.NewPluginLoader("")?
 
What about https://github.com/ipfs/go-ipfs/pull/6474? You'd run:

```go
        l, err := loader.NewPluginLoader()
	if err != nil {
		return nil, err
	}

	err = l.Load(myPlugin) // magic here.
	if err != nil {
		return nil, err
	}

	err = l.Initialize()
	if err != nil {
		return nil, err
	}
	err = l.Inject()
	if err != nil {
		return nil, err
	}
	node, err := ipfs.InitIPFSNode(ipfsPath)
	if err != nil {
		return nil, err
	}
```

**Comment by i-norden** [comment]
Thanks @Stebalien, that works! With that in mind, is there anything else blocking merging this work?

**Pull request merged**